### PR TITLE
IGNITE-18593 Get rid of MvPartitionStorage and TxStateStorage in PartitionAccess

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-18593

### CODE REVIEW

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionAccess.java [hunk 1]
```diff
@@ -32,27 +41,111 @@ public interface PartitionAccess {
...
+     *
+     * @return Future that will complete when the partition is recreated.
+     * @throws StorageException If an error has occurred during the partition destruction.
      */
-    MvPartitionStorage mvPartitionStorage();
+    CompletableFuture<Void> reCreateMvPartitionStorage() throws StorageException;
 
     /**
-     * Returns transaction state storage for the partition.
+     * Destroys and recreates the multi-versioned partition storage.
```
Comment (rpuch) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionAccess.java:52]:
```suggestion
     * Destroys and recreates the TX state partition storage.
```

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionAccess.java [hunk 2]
```diff
@@ -32,27 +41,111 @@ public interface PartitionAccess {
...
+     *
+     * @throws StorageException If an error has occurred during transaction state storage for the partition destruction.
```
Comment (rpuch) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionAccess.java:54]:
```suggestion
     * @throws StorageException If an error has occurred during destruction of the transaction state storage for the partition.
```

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/incoming/IncomingSnapshotCopier.java [hunk 3]
```diff
@@ -186,32 +186,13 @@ public SnapshotReader getReader() {
...
-                        mvPartitionStorage.lastApplied(FULL_RABALANCING_STARTED, 0);
-
-                        return null;
-                    });
-
-                    LOG.info("Copier prepared multi-versioned storage for the partition [partId={}, tableId={}]", partId(), tableId());
-
-                    return completedFuture(null);
-                }, executor);
+        return partitionSnapshotStorage.partition().reCreateMvPartitionStorage();
```
Comment (rpuch) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/incoming/IncomingSnapshotCopier.java:195]: The previous version of the code logged when it prepared the MV storage. Here, the logging is removed. Was this made on purpose, or should we still log it?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/incoming/IncomingSnapshotCopier.java [hunk 4]
```diff
@@ -226,18 +207,7 @@ private CompletableFuture<?> prepareTxStatePartitionStorageForRebalance(Executor
...
 
-        return CompletableFuture.supplyAsync(() -> partitionSnapshotStorage.partition().reCreateTxStatePartitionStorage(), executor)
-                .thenCompose(txStatePartitionStorage -> {
-                    if (canceled) {
-                        return completedFuture(null);
-                    }
-
-                    txStatePartitionStorage.lastApplied(FULL_RABALANCING_STARTED, 0);
-
-                    LOG.info("Copier prepared transaction state storage for the partition [partId={}, tableId={}]", partId(), tableId());
```
Comment (rpuch) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/incoming/IncomingSnapshotCopier.java:237]: Same thing: don't we want to log this anymore?

File: modules/table/src/test/java/org/apache/ignite/internal/table/distributed/raft/snapshot/outgoing/OutgoingSnapshotCommonTest.java [hunk 5]
```diff
@@ -76,12 +64,13 @@ void returnsKeyFromStorage() {
 
     @Test
     void sendsSnapshotMeta() {
-        when(mvPartitionStorage.lastAppliedIndex()).thenReturn(100L);
-        lenient().when(mvPartitionStorage.lastAppliedTerm()).thenReturn(3L);
-        when(txStateStorage.lastAppliedIndex()).thenReturn(100L);
-        lenient().when(txStateStorage.lastAppliedTerm()).thenReturn(3L);
+        lenient().when(partitionAccess.minLastAppliedIndex()).thenReturn(90L);
```
Comment (rpuch) [inline modules/table/src/test/java/org/apache/ignite/internal/table/distributed/raft/snapshot/outgoing/OutgoingSnapshotCommonTest.java:67]: These `lenient()`s seem redundant and should be dropped

File: modules/table/src/test/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionSnapshotStorageFactoryTest.java [hunk 6]
```diff
@@ -44,24 +39,18 @@
...
-    void configureMocks() {
-        when(partitionAccess.mvPartitionStorage()).thenReturn(mvPartitionStorage);
-        when(partitionAccess.txStatePartitionStorage()).thenReturn(txStateStorage);
-    }
-
     @Test
     void testForChoosingMinimumAppliedIndexForMeta() {
-        mvPartitionStorage.lastApplied(10L, 2L);
-        txStateStorage.lastApplied(5L, 1L);
+        lenient().when(partitionAccess.minLastAppliedIndex()).thenReturn(5L);
```
Comment (rpuch) [inline modules/table/src/test/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionSnapshotStorageFactoryTest.java:47]: These `lenient()`s seem redundant (as the corresponding methods are always called under this test), so they should be dropped.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionAccessImpl.java [hunk 7]
```diff
@@ -57,48 +70,147 @@ public PartitionKey partitionKey() {
...
+            txStateStorage.lastApplied(lastAppliedIndex, lastAppliedTerm);
+
+            mvPartitionStorage.committedGroupConfiguration(raftGroupConfig);
+
+            return null;
+        });
+    }
+
+    @Override
+    public long minLastAppliedIndex() {
```
Comment (rpuch) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionAccessImpl.java:170]: Please add unit tests for all 4 min/max methods. This logic was implemented in other places before this change, and it was covered by tests (as it seems really important). So we probably need these tests here now.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionAccess.java (hunk 1 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionAccess.java:52]: Fix it

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionAccess.java (hunk 2 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionAccess.java:54]: Fix it

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/incoming/IncomingSnapshotCopier.java (hunk 4 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/incoming/IncomingSnapshotCopier.java:237]: This code will be removed soon, left TODO

File: modules/table/src/test/java/org/apache/ignite/internal/table/distributed/raft/snapshot/outgoing/OutgoingSnapshotCommonTest.java (hunk 5 above)
Comment (tkalkirill) [reply modules/table/src/test/java/org/apache/ignite/internal/table/distributed/raft/snapshot/outgoing/OutgoingSnapshotCommonTest.java:67]: I tried to delete it, but got the following error: `Please remove unnecessary stubbings or use 'lenient' strictness. More info: javadoc for UnnecessaryStubbingException class.`

File: modules/table/src/test/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionSnapshotStorageFactoryTest.java (hunk 6 above)
Comment (tkalkirill) [reply modules/table/src/test/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionSnapshotStorageFactoryTest.java:47]: I tried to delete it, but got the following error: `Please remove unnecessary stubbings or use 'lenient' strictness. More info: javadoc for UnnecessaryStubbingException class.`

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionAccessImpl.java (hunk 7 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionAccessImpl.java:170]: I think there are already tests:
- `org.apache.ignite.internal.table.distributed.raft.snapshot.PartitionSnapshotStorageFactoryTest#testForChoosingMinimumAppliedIndexForMeta`
- `org.apache.ignite.internal.table.distributed.raft.snapshot.outgoing.OutgoingSnapshotReaderTest#testForChoosingMaximumAppliedIndexForMeta`

File: modules/table/src/test/java/org/apache/ignite/internal/table/distributed/raft/snapshot/outgoing/OutgoingSnapshotCommonTest.java (hunk 5 above)
Comment (rpuch) [reply modules/table/src/test/java/org/apache/ignite/internal/table/distributed/raft/snapshot/outgoing/OutgoingSnapshotCommonTest.java:67]: Then how about removing the whole line? If Mockito says the stubbing is not used, then we probably don't need it at all, right?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionAccessImpl.java (hunk 7 above)
Comment (rpuch) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionAccessImpl.java:170]: Both tests that you mention make sure that `PartitionAccess#min/max...` methods get called. But here, we need to make that that these methods actually work correctly (meaning, that they actually calculate min or max of the corresponding values). This is not tested anywhere.

File: modules/table/src/test/java/org/apache/ignite/internal/table/distributed/raft/snapshot/outgoing/OutgoingSnapshotCommonTest.java (hunk 5 above)
Comment (tkalkirill) [reply modules/table/src/test/java/org/apache/ignite/internal/table/distributed/raft/snapshot/outgoing/OutgoingSnapshotCommonTest.java:67]: Fix it

File: modules/table/src/test/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionSnapshotStorageFactoryTest.java (hunk 6 above)
Comment (tkalkirill) [reply modules/table/src/test/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionSnapshotStorageFactoryTest.java:47]: Fix it

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionAccessImpl.java (hunk 7 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/PartitionAccessImpl.java:170]: Understood you, I will add tests.

**Pull request merged**